retry.NewExponentialWithJitter: Panics when retrying too often

My recent refactoring of `retry.NewExponentialWithJitter` in 86113e96795a657183b99a5e0e38b768bb349dec introduced the possibility for an overflow if a retry happens lots of times.

The following line may result in `int64(min << attempt)` to become negative.

https://github.com/Icinga/icinga-go-library/blob/003f19a6a8b38bedf4bf8989ea9639bb096c2ca4/backoff/backoff.go#L30

And now the initial check will be mitigated.

https://github.com/Icinga/icinga-go-library/blob/003f19a6a8b38bedf4bf8989ea9639bb096c2ca4/backoff/backoff.go#L43-L49

Resulting in a panic.

```
panic: recovered from errgroup.Group: invalid argument to Int63n
        goroutine 13 [running]:
        runtime/debug.Stack()
                /nix/store/5xvi25nqmbrg58aixp4zgczilfnp7pwg-go-1.24.3/share/go/src/runtime/debug/stack.go:26 +0x5e
        golang.org/x/sync/errgroup.(*Group).add.func1.1()
                /home/apenning/go/pkg/mod/golang.org/x/sync@v0.14.0/errgroup/errgroup.go:124 +0x139
        panic({0xbb7d80?, 0xda83d0?})
                /nix/store/5xvi25nqmbrg58aixp4zgczilfnp7pwg-go-1.24.3/share/go/src/runtime/panic.go:792 +0x132
        math/rand.(*Rand).Int63n(0x0?, 0xc21720?)
                /nix/store/5xvi25nqmbrg58aixp4zgczilfnp7pwg-go-1.24.3/share/go/src/math/rand/rand.go:122 +0xcb
        math/rand.Int63n(0xfa12000000000000)
                /nix/store/5xvi25nqmbrg58aixp4zgczilfnp7pwg-go-1.24.3/share/go/src/math/rand/rand.go:454 +0x25
        github.com/icinga/icinga-go-library/backoff.jitter(0x0?)
                /home/apenning/go/pkg/mod/github.com/icinga/icinga-go-library@v0.7.0/backoff/backoff.go:48 +0x32
        github.com/icinga/icinga-go-library/backoff.init.NewExponentialWithJitter.func1(0x191c2dc0b9f?)
                /home/apenning/go/pkg/mod/github.com/icinga/icinga-go-library@v0.7.0/backoff/backoff.go:30 +0x35
        github.com/icinga/icinga-go-library/retry.WithBackoff({0xdaeef0, 0xc0000b82d0}, 0xc000fd37c8, 0xce7190, 0xc000140000, {0x547fa5?, 0xc00
0ebf818?, 0xc000ebf800?})
                /home/apenning/go/pkg/mod/github.com/icinga/icinga-go-library@v0.7.0/retry/retry.go:110 +0x277
        github.com/icinga/icinga-go-library/redis.NewClientFromConfig.dialWithLogging.func1({0xdaeef0, 0xc0000b82d0}, {0xcb94b4, 0x3}, {0xc0002
8c5b0, 0xe})
                /home/apenning/go/pkg/mod/github.com/icinga/icinga-go-library@v0.7.0/redis/client.go:313 +0x17f
        github.com/redis/go-redis/v9.(*baseClient).dial(...)
```

A small [demo](https://go.dev/play/p/QH-SS9VqiwB) shows that this happens at the 37th attempt when using a `min` of 128ms, the default for `retry.DefaultBackoff`.

--- a/backoff/backoff.go
+++ b/backoff/backoff.go
@@ -27,7 +27,15 @@ func NewExponentialWithJitter(min, max time.Duration) Backoff {
 	}
 
 	return func(attempt uint64) time.Duration {
-		e := time.Duration(jitter(int64(min << attempt)))
+		e := min << attempt
+		// If the bit shift already overflows, return max.
+		if e < min {
+			return max
+		}
+
+		// Introduce jitter. e <- [min/2, int64_max)
+		e = e/2 + time.Duration(rand.Int63n(int64(e/2))) // #nosec G404 -- we don't need crypto/rand here though.
+		// Remap e to [min, max].
 		if e < min {
 			e = min
 		}
@@ -37,13 +45,4 @@ func NewExponentialWithJitter(min, max time.Duration) Backoff {
 
 		return e
 	}
-}
-
-// jitter returns a random integer distributed in the range [n/2..n).
-func jitter(n int64) int64 {
-	if n == 0 {
-		return 0
-	}
-
-	return n/2 + rand.Int63n(n/2) // #nosec G404 -- Use of weak random number generator - we don't need crypto/rand here though.
 }

--- a/backoff/backoff_test.go
+++ b/backoff/backoff_test.go
@@ -2,6 +2,7 @@ package backoff
 
 import (
 	"github.com/stretchr/testify/require"
+	"math"
 	"testing"
 	"time"
 )
@@ -22,25 +23,23 @@ func TestNewExponentialWithJitter(t *testing.T) {
 		t.Run(tt.name, func(t *testing.T) {
 			r := NewExponentialWithJitter(tt.min, tt.max)
 
+			require.Equal(t, tt.min, r(0))
+			require.Equal(t, tt.max, r(math.MaxUint64))
+
 			// Ensure that multiple calls don't breach the upper bound
-			maxCounter := 0
+			reachedMax := false
 
-			for i := uint64(0); ; i++ {
-				if maxCounter >= 10 {
-					break
-				}
-				if i > 1_000_000 {
-					t.Error("not reached max")
-				}
-
+			for i := uint64(0); i < 1024; i++ {
 				d := r(i)
 				require.GreaterOrEqual(t, d, tt.min)
 				require.LessOrEqual(t, d, tt.max)
 
+				require.Falsef(t, reachedMax && d != tt.max, "max value %v was already reached, but r(%d) := %v", tt.max, i, d)
 				if d == tt.max {
-					maxCounter++
+					reachedMax = true
 				}
 			}
+			require.True(t, reachedMax, "max value was never reached")
 		})
 	}
 }

